stdlib/path.gr

```diff
@@ -0,0 +1,775 @@
+    [Dot | Dotdot | Slash, ..._] => false,
+    _ => true,
+  }
+}
+
+let parseAbs = (pathStr, fileType, platform) => {
+  match (lexPath(pathStr, platform)) {
+    [] => None,
+    [first, ...rest] => {
+      let tokens = [first, ...rest]
+      if (fileType == File && !isFilePath(tokens)) {
+        None
+      } else {
+        let init = match (first) {
+          Slash => Some((Abs(Root), fileType, [])),
+          DriveTok(label) => Some((Abs(Drive(label)), fileType, [])),
+          _ => None,
+        }
+        Option.map(init => List.reduce(parseNextToken, init, rest), init)
+      }
+    },
+  }
+}
+
+// utility functions to translate path types
+
+let toTyped = (pathInfo: PathInfo) => {
+  let (base, fileType, subs) = pathInfo
+  ({ base, }, { fileType, }, subs): TypedPath<a, b>
+}
+
+let toUntyped = (typedPath: TypedPath<a, b>) => {
+  let (base, fileType, subs) = typedPath
+  let { base } = base
+  let { fileType } = fileType
+  (base, fileType, subs): PathInfo
+}
+
+let pathInfo = (path: Path) => {
+  match (path) {
+    AbsoluteDir(x) => toUntyped(x),
+    AbsoluteFile(x) => toUntyped(x),
+    RelativeDir(x) => toUntyped(x),
+    RelativeFile(x) => toUntyped(x),
+  }
+}
+
+let toPath = (path: PathInfo) => {
+  match (path) {
+    (Abs(_), File, _) as p => AbsoluteFile(toTyped(p)),
+    (Abs(_), Dir, _) as p => AbsoluteDir(toTyped(p)),
+    (Rel(_), File, _) as p => RelativeFile(toTyped(p)),
+    (Rel(_), Dir, _) as p => RelativeDir(toTyped(p)),
+  }
+}
+
+// TODO(#1496): expose these functions once module system added
+
+let absoluteFile = (pathStr, platform) => {
+  let pathOpt = parseAbs(pathStr, File, platform)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Absolute, File>>
+}
+
+let absoluteDir = (pathStr, platform) => {
+  let pathOpt = parseAbs(pathStr, Dir, platform)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Absolute, Directory>>
+}
+
+let parseRel = (pathStr, fileType, platform) => {
+  let (first, rest) = match (lexPath(pathStr, platform)) {
+    [] => (Dot, []),
+    [first, ...rest] => (first, rest),
+  }
+  let tokens = [first, ...rest]
+
+  if (fileType == File && !isFilePath(tokens)) {
+    None
+  } else {
+    let init = match (first) {
+      Dot => Some((Rel(0), fileType, [])),
+      Dotdot => Some((Rel(1), fileType, [])),
+      Text(str) => Some((Rel(0), fileType, [str])),
+      _ => None,
+    }
+    Option.map(init => List.reduce(parseNextToken, init, rest), init)
+  }
+}
+
+let relativeFile = (pathStr, platform) => {
+  let pathOpt = parseRel(pathStr, File, platform)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Relative, File>>
+}
+
+let relativeDir = (pathStr, platform) => {
+  let pathOpt = parseRel(pathStr, Dir, platform)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Relative, Directory>>
+}
+
+// TODO(#1496): reuse this and the other helper functions for the typed implementations
+let fromStringHelper = (pathStr, platform) => {
+  let tokens = match (lexPath(pathStr, platform)) {
```

Comment:
I'm noticing that we actually lex the path twice for this API. I wonder if the `parse` functions should take a token list instead of the `pathStr`. Thoughts?

Reply:
Yeah good idea, I was actually planning on doing this at some point but forgot to implement it.